packages/artifact/src/internal/download-http-client.ts

```diff
@@ -124,66 +168,128 @@ export class DownloadHttpClient {
       )
     }
 
-    if (isSuccessStatusCode(response.message.statusCode)) {
-      await this.pipeResponseToStream(
-        response,
-        stream,
-        isGzip(response.message.headers)
-      )
-    } else if (isRetryableStatusCode(response.message.statusCode)) {
-      warning(
-        `Received http ${response.message.statusCode} during file download, will retry ${artifactLocation} after 10 seconds`
-      )
-      // if an error is encountered, dispose of the http connection, and create a new one
-      this.downloadHttpManager.disposeAndReplaceClient(httpClientIndex)
-      await new Promise(resolve =>
-        setTimeout(resolve, getRetryWaitTimeInMilliseconds())
-      )
-      const retryResponse = await client.get(artifactLocation)
-      if (isSuccessStatusCode(retryResponse.message.statusCode)) {
-        await this.pipeResponseToStream(
+    // Increments the current retry count and then checks if the retry limit has been reached
+    // If there have been too many retries, fail so the download stops. If there is a retryAfterValue value provided,
+    // it will be used
+    const backOff = async (retryAfterValue?: number): Promise<void> => {
+      retryCount++
+      if (retryCount > retryLimit) {
+        return Promise.reject(
+          new Error(
+            `Retry limit has been reached. Unable to download ${artifactLocation}`
+          )
+        )
+      } else {
+        this.downloadHttpManager.disposeAndReplaceClient(httpClientIndex)
+        if (retryAfterValue) {
+          // Back off exponentially based off of the retry count
+          core.info(
+            `Backoff due to too many requests, retry #${retryCount}. Waiting for ${retryAfterValue} milliseconds before continuing the download`
+          )
+          await new Promise(resolve => setTimeout(resolve, retryAfterValue))
+        } else {
+          // Back off using an exponential value that depends on the retry count
+          const backoffTime = getExponentialRetryTimeInMilliseconds(retryCount)
+          core.info(
+            `Exponential backoff for retry #${retryCount}. Waiting for ${backoffTime} milliseconds before continuing the download`
+          )
+          await new Promise(resolve => setTimeout(resolve, backoffTime))
+        }
+        core.info(
+          `Finished backoff for retry #${retryCount}, continuing with download`
+        )
+      }
+    }
+
+    // keep trying to download a file until a retry limit has been reached
+    while (retryCount <= retryLimit) {
+      let response: IHttpClientResponse
+      try {
+        response = await makeDownloadRequest()
+      } catch (error) {
+        // if an error is caught, it is usually indicative of a timeout so retry the download
+        core.info(
+          'An error has been caught, while attempting to download a file'
```

Comment:
Nit: `caught` doesn't feel right in an error message to the user.  Maybe `An error occurred while attempting to download a file`?